In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/carrie1/ecommerce-data/data.csv


In [2]:
df=pd.read_csv('/kaggle/input/datasets/carrie1/ecommerce-data/data.csv',encoding='latin1')
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [3]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [4]:
df['InvoiceNo'].duplicated().sum()

np.int64(516009)

In [5]:
df.duplicated().sum()

np.int64(5268)

In [6]:
df[df.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,12/1/2010 11:45,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,12/1/2010 11:45,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,12/1/2010 11:45,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,12/1/2010 11:45,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,12/1/2010 11:49,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,12/9/2011 11:34,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,12/9/2011 11:34,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,12/9/2011 11:34,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,12/9/2011 11:34,2.10,14446.0,United Kingdom


In [7]:
(
    df
    .groupby(df.columns.tolist())
    .size()
    .reset_index(name='conteo')
    .query('conteo > 1')
)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,conteo
491,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,12/1/2010 11:45,1.25,17908.0,United Kingdom,2
499,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,12/1/2010 11:45,4.95,17908.0,United Kingdom,2
510,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,12/1/2010 11:45,2.10,17908.0,United Kingdom,2
511,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,12/1/2010 11:45,2.95,17908.0,United Kingdom,2
537,536412,21448,12 DAISY PEGS IN WOOD BOX,1,12/1/2010 11:49,1.65,17920.0,United Kingdom,2
...,...,...,...,...,...,...,...,...,...
399976,C572226,85066,CREAM SWEETHEART MINI CHEST,-1,10/21/2011 13:58,12.75,15321.0,United Kingdom,2
400277,C574095,22326,ROUND SNACK BOXES SET OF4 WOODLAND,-1,11/3/2011 9:54,2.95,12674.0,France,2
400365,C574510,22360,GLASS JAR ENGLISH CONFECTIONERY,-1,11/4/2011 13:25,2.95,15110.0,United Kingdom,2
400646,C575940,23309,SET OF 60 I LOVE LONDON CAKE CASES,-24,11/13/2011 11:38,0.55,17838.0,United Kingdom,2


Se detectaron 5268 registros completamente duplicados tras el profiling inicial del dataset. Después de validar muestras manuales, los registros fueron eliminados para evitar inflación de métricas de ventas

In [8]:
df = df.drop_duplicates()

In [9]:
df[df['Quantity'] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


In [10]:
df[df['UnitPrice'] == 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,12/8/2011 13:58,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,12/8/2011 13:58,0.0,NaN,United Kingdom
538554,581408,85175,NaN,20,12/8/2011 14:06,0.0,NaN,United Kingdom


In [11]:
df_canelation=df[df['InvoiceNo'].astype(str).str.startswith('C')]
df_negative=df[df['Quantity'] < 0]

df_canelation

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


In [12]:
df_canelation.groupby('StockCode')['Description'].count()

StockCode
10133      1
10135      1
11001      5
15034      4
15036      4
        ... 
D         77
DOT        1
M        244
POST     126
S         60
Name: Description, Length: 1947, dtype: int64

In [13]:
productos_negative = set(df_negative['StockCode'])
productos_cancelation = set(df_canelation['StockCode'])

anomalias = productos_negative - productos_cancelation



In [14]:
df_anomalias=df_negative[df_negative['StockCode'].isin(anomalias)][['StockCode', 'Description']].drop_duplicates()
df_anomalias.head()

,StockCode,Description
4347,84952C,NaN
7190,85067,NaN
7192,21414,NaN
7193,21653,NaN
7196,21814,NaN


In [15]:
price_zero = df[df['UnitPrice'] == 0]

price_zero[['StockCode', 'Description']].drop_duplicates()

price_negative = df[df['UnitPrice'] < 0]

price_negative[['StockCode', 'Description']].drop_duplicates()



,StockCode,Description
299983,B,Adjust bad debt


In [16]:
cancelaciones = df[df['InvoiceNo'].astype(str).str.startswith('C')]

productos_cancelados = set(cancelaciones['StockCode'])
productos_price_negative = set(price_negative['StockCode'])

anomalias_price = productos_price_negative - productos_cancelados



In [17]:
# =========================
# 1. Eliminar duplicados exactos
# =========================
df_clean = df.drop_duplicates().copy()

# =========================
# 2. Separar cancelaciones válidas
# =========================
df_cancel = df_clean[
    df_clean['InvoiceNo'].astype(str).str.startswith('C')
]

# =========================
# 3. Negativos anómalos
# Quantity < 0 SIN cancelación
# =========================
df_negativos_anomalos = df_clean[
    (df_clean['Quantity'] < 0) &
    (~df_clean['InvoiceNo'].astype(str).str.startswith('C'))
]

# =========================
# 4. Precios negativos anómalos
# UnitPrice < 0 SIN cancelación
# =========================
df_price_anomalos = df_clean[
    (df_clean['UnitPrice'] < 0) &
    (~df_clean['InvoiceNo'].astype(str).str.startswith('C'))
]

# =========================
# 5. Eliminar anomalías
# =========================
indices_remove = set(df_negativos_anomalos.index).union(
    set(df_price_anomalos.index)
)

df_final = df_clean.drop(index=indices_remove).copy()

# =========================
# 6. Etiquetar precios cero
# =========================
df_final['price_zero_flag'] = df_final['UnitPrice'] == 0

# =========================
# 7. Revenue
# =========================
df_final['Revenue'] = (
    df_final['Quantity'] * df_final['UnitPrice']
)

# =========================
# Dataset final
# =========================
df_final

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,price_zero_flag,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom,False,15.30
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom,False,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom,False,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom,False,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom,False,20.34
...,...,...,...,...,...,...,...,...,...,...
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France,False,10.20
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France,False,12.60
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France,False,16.60
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France,False,16.60


In [18]:
# Revenue por país
df_final.groupby('Country')['Revenue'].sum()

Country
Australia                137009.770
Austria                   10154.320
Bahrain                     548.400
Belgium                   40910.960
Brazil                     1143.600
Canada                     3666.380
Channel Islands           20076.390
Cyprus                    12858.760
Czech Republic              707.720
Denmark                   18768.140
EIRE                     262993.380
European Community         1291.750
Finland                   22326.740
France                   197317.110
Germany                  221509.470
Greece                     4710.520
Hong Kong                  9908.240
Iceland                    4310.000
Israel                     7901.970
Italy                     16890.510
Japan                     35340.620
Lebanon                    1693.880
Lithuania                  1661.060
Malta                      2505.470
Netherlands              284661.540
Norway                    35163.460
Poland                     7213.140
Portugal            

In [19]:
# Top productos
df_final['Description'].value_counts(ascending=False)

Description
WHITE HANGING HEART T-LIGHT HOLDER     2357
REGENCY CAKESTAND 3 TIER               2189
JUMBO BAG RED RETROSPOT                2156
PARTY BUNTING                          1720
LUNCH BAG RED RETROSPOT                1625
                                       ... 
SET 10 CARDS SNOWY ROBIN 17099            1
SET 10 CARDS SWIRLY XMAS TREE 17104       1
LETTER "U" BLING KEY RING                 1
check?                                    1
dotcom adjust                             1
Name: count, Length: 4093, dtype: int64

In [20]:
df_final.groupby('Description')['Quantity'].sum()
df_final.groupby('Description')['Revenue'].sum()

Description
 4 PURPLE FLOCK DINNER CANDLES     285.70
 50'S CHRISTMAS GIFT BAG LARGE    2341.13
 DOLLY GIRL BEAKER                2877.50
 I LOVE LONDON MINI BACKPACK      1624.02
 I LOVE LONDON MINI RUCKSACK         4.15
                                   ...   
wrongly coded 20713                  0.00
wrongly coded 23343                  0.00
wrongly marked                       0.00
wrongly marked 23343                 0.00
wrongly sold (22719) barcode         0.00
Name: Revenue, Length: 4093, dtype: float64

In [21]:
# Clientes más valiosos

df_final.groupby('CustomerID')['Revenue'].sum()

CustomerID
12346.0       0.00
12347.0    4310.00
12348.0    1797.24
12349.0    1757.55
12350.0     334.40
            ...   
18280.0     180.60
18281.0      80.82
18282.0     176.60
18283.0    2045.53
18287.0    1837.28
Name: Revenue, Length: 4372, dtype: float64